text/0023-disconnect-from-offenders.md

```diff
@@ -0,0 +1,91 @@
+
+There is also a question of what is the adequate response to an offence for each of these 3 points. This CEP proposes that for offences not constituting a threat to the security of the network (eg., invalid messages that we can just detect and discard, as opposed to something like equivocations, which can cause problems to consensus), we shouldn't even bother telling other nodes - just disconnecting and blacklisting should be enough. More on that in the [Rationale and alternatives](#rationale-and-alternatives) section.
+
+Thus, we would define a new mechanism: a `DisconnectAndBlacklist` request to the networking component (see [unresolved question 1](#unresolved-questions)).
+
+The `DisconnectAndBlacklist` request would cause the networking component to disconnect from the given node and add it to a blacklist, preventing it from reconnecting in the future. Specifically, it would blacklist the node's endpoint (IP and port). In order to keep the size of the blacklist manageable, we could consider various kinds of compression mechanisms and simplifications, such as blacklisting a whole IP address (all ports) if the same address shows up multiple times.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+We would add the following to `effect::requests::NetworkRequest`:
+
+```rust
+enum NetworkRequest<I, P> {
+    // ...other variants
+    DisconnectAndBlacklist { node: I },
+}
+```
+
+The networking component would also keep a blacklist of nodes' endpoints (see [unresolved question 2](#unresolved-questions)). If a node attempts to connect from a blacklisted endpoint, the connection will be rejected (see [unresolved question 3](#unresolved-questions)).
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+By choosing not to accuse nodes sending invalid messages to other nodes, we theoretically make it possible for a known malicious node to participate in the network for a while. However, the risk involved is rather small (see [Rationale and alternatives](#rationale-and-alternatives) below).
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Most of the rationale behind the proposal has been explained [above](#motivation).
+
+One missing part is why not accuse offending nodes. The issue here is that accusation are a nonzero, even if small, cost to the network. A determined malicious party could trick the network into spamming itself with accusations by starting many nodes sending invalid messages. On the other hand, the risk associated with not immediately banning such a node is small. If an offender only sends valid messages to other nodes, there is no harm in still letting it communicate with the network, and if it tries sending an invalid one, it will just be blacklisted by the recipient independently. This means that we wouldn't be gaining much by accusing such nodes, but could potentially introduce a vector for a DoS-type attack.
```

Comment:
You could mention that creating new node identity (keys) is cheap so banning a key is not a solution.